plugins/speccy.py

```diff
@@ -0,0 +1,135 @@
+# -*- coding: utf-8 -*-
+from cloudbot import hook
+from cloudbot.event import EventType
+
+import asyncio
+import re
+import requests
+import tempfile
+
+from bs4 import BeautifulSoup as BS
+
+
+@asyncio.coroutine
+@hook.event([EventType.message, EventType.action], singlethread=True)
+def get_speccy_url(conn, message, chan, content, nick):
+    re_content = re.search(
+        r"https?:\/\/speccy.piriform.com\/results\/[A-z0-9]+", content)
+    if re_content:
+        return parse_speccy(message, nick, str(re_content.group(0)))
+
+
+def parse_speccy(message, nick, url):
+
+    response = requests.get(url)
+    if not response:
+        return None
+
+    respHtml = response.content
+    speccy = tempfile.NamedTemporaryFile()
+
+    with open(speccy.name, 'wb') as f:
+        f.write(respHtml)
+
+    soup = BS(open(speccy.name), "lxml-xml")
```

Comment:
Going to look into the other two requests, but about this one. This was actually an optimization improvement. If you know a better way lmk. 

https://i.imgur.com/G2Si7PP.png

I noticed it also parses the link faster too:

https://i.imgur.com/R1eKAZT.png

test.py = the method you mentioned test2.py = the current method seen in snippet

Reply:
Please include your test case. Also the `time` command shows system resource usage not actual elapsed time. Disk I/O has a lot of waiting since disk I/O is slow. This code is also running in the main thread as you have marked it as a coroutine, so any blockling I/O like an HTTP request or file write will halt the entire bot until it finishes.

Reply:
I don't exactly know why it's faster, it was just a guess that ended up working, but here so you can try yourself:

https://gist.github.com/Jessexd/14990cbadad5e62a22fb590bb25a4f9d

Should I add multithreaded=True ?

Reply:
No, you should remove the `@asyncio.coroutine` decorator on your hook function, which is marking it as a coroutine, which runs in the main event loop.

Reply:
Ah, alright, as for BS, what should be done in terms of parsing links faster instead of writing to disk?

Reply:
Generally just pass the content from the response straight in.
```python
import requests

from bs4 import BeautifulSoup

with requests.get(url) as response:
    soup = BeautifulSoup(response.content, 'lxml', from_encoding=response.encoding)

print(soup)
```

It may not be the fastest option, but it's the standard way as something like disk i/o can be extremely unpredictable.